### PR TITLE
chore: update Dockefile for nodegyp support downstream; remove all the needless named intermediate stages that don't work in OSBS

### PR DESCRIPTION
**Pull request by nickboldt**
### What does this PR do?

chore: update Dockefile for nodegyp support downstream; remove all the needless named intermediate stages that don't work in OSBS

Signed-off-by: Nick Boldt <nboldt@redhat.com>

### Screenshot/screencast of this PR
N/A

### What issues does this PR fix or reference?
N/A (or see commit message above for issue number)

### How to test this PR?
N/A

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [x] [The repository devfile is up to date and works, if one exists](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections What issues does this PR fix or reference and How to test this PR completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [x] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -20,87 +20,146 @@
...
 # Downstream comment
 ENV EXTERNAL_SOURCE=.
+ENV EXTERNAL_SOURCE_NESTED=.
 ENV CONTAINER_SOURCE=/opt/app-root/src
 #/ Downstream comment
-
 # Downstream sources
 # Downstream uncomment
 # ENV EXTERNAL_SOURCE=$REMOTE_SOURCES/upstream1/app
+# ENV EXTERNAL_SOURCE_NESTED=$REMOTE_SOURCES/upstream1/app/distgit/containers/rhdh-hub
```
Comment (schultzp2020) [inline Dockerfile:53]:
```suggestion
# ENV EXTERNAL_SOURCE_NESTED=$EXTERNAL_SOURCE/distgit/containers/rhdh-hub
```

**Review by schultzp2020 — CHANGES_REQUESTED** [review]

**Comment by schultzp2020** [comment]
@christophe-f should we create another Dockerfile for the downstream since they are diverging a decent bit.

File: Dockerfile [hunk 2]
```diff
@@ -20,87 +20,146 @@
...
-#/ Downstream uncomment
+# Downstream comment
+# RUN rm app-config.yaml && mv app-config.example.yaml app-config.yaml
+#/ Downstream comment
 
+# hadolint ignore=DL3059
 RUN $YARN build --filter=backend
 
 # Stage 4 - Build the actual backend image and install production dependencies
+# Downstream comment
```
Comment (nickboldt) [inline Dockerfile:136]: if you prefer, I can just skip these unsupported intermediate stages with a sed command -- https://gitlab.cee.redhat.com/rhidp/cpaas-rhdh-hub/-/blob/rhdh-1.0-rhel-9/sync.sh#L242-243

File: Dockerfile [hunk 3]
```diff
@@ -20,87 +20,146 @@
...
-COPY $EXTERNAL_SOURCE/packages/app/package.json ./packages/app/package.json
-COPY $EXTERNAL_SOURCE/packages/backend/package.json ./packages/backend/package.json
+COPY $EXTERNAL_SOURCE_NESTED/package.json $EXTERNAL_SOURCE_NESTED/yarn.lock ./
+COPY $EXTERNAL_SOURCE_NESTED/packages/app/package.json ./packages/app/package.json
+COPY $EXTERNAL_SOURCE_NESTED/packages/backend/package.json ./packages/backend/package.json
 
 # see https://docs.engineering.redhat.com/pages/viewpage.action?pageId=228017926#UpstreamSources(Cachito,ContainerFirst)-CachitoIntegrationfornpm
 # Downstream uncomment
-# RUN source $CONTAINER_SOURCE/cachito.env && $YARN config set nodedir /usr; $YARN config set unsafe-perm true; # $YARN config list --verbose
+# COPY $REMOTE_SOURCES/upstream1/cachito.env \
```
Comment (nickboldt) [inline Dockerfile:84]: this whole block can be made smaller if upstream prefers the debugging stuff removed and streamlined.

**Comment by schultzp2020** [comment]
> if you prefer, I can just skip these unsupported intermediate stages with a sed command

> this whole block can be made smaller if upstream prefers the debugging stuff removed and streamlined.

@nickboldt both of these are probably a good idea. Ideally we should try to hide as much of the downstream as possible to make it easier to understand.

**Comment by nickboldt** [comment]
> > if you prefer, I can just skip these unsupported intermediate stages with a sed command
> 
> > this whole block can be made smaller if upstream prefers the debugging stuff removed and streamlined.
> 
> @nickboldt both of these are probably a good idea. Ideally we should try to hide as much of the downstream as possible to make it easier to understand.

While I would prefer to say https://imgflip.com/i/7wjiiu?herp=1692707629737 I can rework this a bit to move some changes to the midstream.

HOWEVER this does mean the next time you change things in the Dockerfile, you'll have to check if you broke midstream.

Simpler might be to have TWO Dockerfiles... one for upstream and one called Dockerfile.rhel9 (?) so you can at least see the changes needed for both files in the same repo

**Comment by christophe-f** [comment]
> @christophe-f should we create another Dockerfile for the downstream since they are diverging a decent bit.

Yes we could try that. We could have 2 files in the same repo, or one in midstream and downstream. We want to keep this simple and easy to maintain. We can also iterate on it and try different things.

**Review by schultzp2020 — APPROVED** [review]
LGTM

**Pull request merged**